File: engines/go/internal/format/format.go
package format

import (
	"fmt"
	"strings"
)

// FmtK formats token counts: 1234567->"1.2M", 45231->"45k", 1234->"1.2k", 523->"523".
func FmtK(n int) string {
	if n >= 1000000 {
		return fmt.Sprintf("%.1fM", float64(n)/1000000)
	}
	if n >= 10000 {
		return fmt.Sprintf("%.0fk", float64(n)/1000)
	}
	if n >= 1000 {
		return fmt.Sprintf("%.1fk", float64(n)/1000)
	}
	return fmt.Sprintf("%d", n)
}

// FmtCost formats cost: >=1000->"$1.8k", >=100->"$374", >=10->"$14", >=1->"$8.4", <1->"$0.12".
func FmtCost(c float64) string {
	if c >= 1000 {
		return fmt.Sprintf("$%.1fk", c/1000)
	}
	if c >= 100 {
		return fmt.Sprintf("$%.0f", c)
	}
	if c >= 10 {
		return fmt.Sprintf("$%.0f", c)
	}
	if c >= 1 {
		return fmt.Sprintf("$%.1f", c)
	}
	return fmt.Sprintf("$%.2f", c)
}

// FmtDuration formats milliseconds: >=60min->"4h0m", <60min->"15m".
func FmtDuration(ms int) string {
	min := ms / 60000
	if min >= 60 {
		return fmt.Sprintf("%dh%dm", min/60, min%60)
	}
	return fmt.Sprintf("%dm", min)
}

var bars = []string{"\u2581", "\u2582", "\u2583", "\u2584", "\u2585", "\u2586", "\u2587", "\u2588"}

// BarChar returns a bar character proportional to val/max (8 levels).
func BarChar(val, max int) string {
	if val <= 0 || max <= 0 {
		return ""
	}
	level := (val*8 + max/2) / max
	if level < 1 {
		level = 1
	}
	if level > 8 {
		level = 8
	}
	return bars[level-1]
}

var branchPrefixes = []struct {
	prefix string
	icon   string
}{
	{"feature/", "\u2605"},
	{"feat/", "\u2605"},
	{"fix/", "\u2726"},
	{"chore/", "\u2699"},
	{"refactor/", "\u21bb"},
	{"docs/", "\u00a7"},
}

// ShortenBranch replaces known prefixes with icons.
func ShortenBranch(name string) string {
	for _, p := range branchPrefixes {
		if strings.HasPrefix(name, p.prefix) {
			return p.icon + name[len(p.prefix):]
		}
	}
	return name
}

// Truncate truncates to maxLen runes with ellipsis.
func Truncate(s string, maxLen int) string {
	runes := []rune(s)
	if len(runes) > maxLen {
		return string(runes[:maxLen-1]) + "\u2026"
	}
	return s
}

File: engines/rust/src/format.rs
/// Format token counts: 1234567->"1.2M", 45231->"45k", 1234->"1.2k", 523->"523".
pub fn fmt_k(n: i64) -> String {
    if n >= 1_000_000 {
        format!("{:.1}M", n as f64 / 1_000_000.0)
    } else if n >= 10_000 {
        format!("{:.0}k", n as f64 / 1_000.0)
    } else if n >= 1_000 {
        format!("{:.1}k", n as f64 / 1_000.0)
    } else {
        format!("{}", n)
    }
}

/// Format cost: >=1000->"$1.8k", >=100->"$374", >=10->"$14", >=1->"$8.4", <1->"$0.12".
pub fn fmt_cost(c: f64) -> String {
    if c >= 1000.0 {
        format!("${:.1}k", c / 1000.0)
    } else if c >= 10.0 {
        format!("${:.0}", c)
    } else if c >= 1.0 {
        format!("${:.1}", c)
    } else {
        format!("${:.2}", c)
    }
}

/// Format milliseconds: >=60min->"4h0m", <60min->"15m".
pub fn fmt_duration(ms: i64) -> String {
    let min = ms / 60_000;
    if min >= 60 {
        format!("{}h{}m", min / 60, min % 60)
    } else {
        format!("{}m", min)
    }
}

const BARS: [&str; 8] = [
    "\u{2581}", "\u{2582}", "\u{2583}", "\u{2584}", "\u{2585}", "\u{2586}", "\u{2587}", "\u{2588}",
];

/// Return a bar character proportional to val/max (8 levels).
pub fn bar_char(val: i64, max: i64) -> &'static str {
    if val <= 0 || max <= 0 {
        return "";
    }
    let level = ((val * 8 + max / 2) / max).clamp(1, 8);
    BARS[(level - 1) as usize]
}

const BRANCH_PREFIXES: &[(&str, &str)] = &[
    ("feature/", "\u{2605}"),
    ("feat/", "\u{2605}"),
    ("fix/", "\u{2726}"),
    ("chore/", "\u{2699}"),
    ("refactor/", "\u{21bb}"),
    ("docs/", "\u{00a7}"),
];

/// Replace known branch prefixes with icons.
pub fn shorten_branch(name: &str) -> String {
    for &(prefix, icon) in BRANCH_PREFIXES {
        if let Some(rest) = name.strip_prefix(prefix) {
            return format!("{}{}", icon, rest);
        }
    }
    name.to_string()
}

/// Truncate to max_len runes with ellipsis.
pub fn truncate(s: &str, max_len: usize) -> String {
    let chars: Vec<char> = s.chars().collect();
    if chars.len() > max_len {
        let mut result: String = chars[..max_len - 1].iter().collect();
        result.push('\u{2026}');
        result
    } else {
        s.to_string()
    }
}

/// Round to even (banker's rounding), matching Go's math.RoundToEven.
pub fn round_to_even(x: f64) -> i64 {
    let rounded = x.round();
    // Check if we're exactly at .5
    if (x - x.floor() - 0.5).abs() < f64::EPSILON {
        // Round to even
        let r = rounded as i64;
        if r % 2 != 0 {
            return r - 1;
        }
    }
    rounded as i64
}
